java/ql/src/experimental/Security/CWE/CWE-522/InsecureBasicAuth.ql

```diff
@@ -0,0 +1,176 @@
+/**
+ * @name Insecure basic authentication
+ * @description Basic authentication only obfuscates username/password in Base64 encoding, which can be easily recognized and reversed. Transmission of sensitive information not over HTTPS is vulnerable to packet sniffing.
+ * @kind path-problem
+ * @id java/insecure-basic-auth
+ * @tags security
+ *       external/cwe-522
+ *       external/cwe-319
+ */
+
+import java
+import semmle.code.java.frameworks.Networking
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * The Java class `org.apache.http.client.methods.HttpRequestBase`. Popular subclasses include `HttpGet`, `HttpPost`, and `HttpPut`.
+ * And the Java class `org.apache.http.message.BasicHttpRequest`.
+ */
+class ApacheHttpRequest extends RefType {
+  ApacheHttpRequest() {
+    this
+        .getASourceSupertype*()
+        .hasQualifiedName("org.apache.http.client.methods", "HttpRequestBase") or
+    this.getASourceSupertype*().hasQualifiedName("org.apache.http.message", "BasicHttpRequest")
+  }
+}
+
+/**
+ * Class of Java URL constructor.
+ */
+class URLConstructor extends ClassInstanceExpr {
+  URLConstructor() { this.getConstructor().getDeclaringType() instanceof TypeUrl }
+
+  Expr stringArg() {
+    // URLs constructed with any of the four string constructors below:
+    // URL(String spec), URL(String protocol, String host, int port, String file), URL(String protocol, String host, int port, String file, URLStreamHandler handler), URL(String protocol, String host, String file)
+    this.getConstructor().getParameter(0).getType() instanceof TypeString and
+    result = this.getArgument(0) // First argument contains the protocol part.
```

Comment:
Note that for the `URL(String protocol, ...)` constructors this could include false positives since the private IP address check is not performed for the host, e.g. `new URL("http", "localhost", "")` would be erroneously flagged.

This could be worked around by having this predicate return `protocol + "://" + host` (`getArgument(0) + "://" + getArgument(1)`) for the `URL(String protocol, ...)` constructors.

Reply:
I've modified the query to check the protocol+host pattern.